internal/db/index.go

```diff
@@ -430,3 +425,323 @@ func isUpdatingIndexedFields(index CollectionIndex, oldDoc, newDoc *client.Docum
+		arrVal := key.Fields[arrFieldIndex].Value
+		normVals, err := client.ToArrayOfNormalValues(arrVal)
+		if err != nil {
+			return nil, err
+		}
+		sets := make(map[client.NormalValue]struct{})
+		for i := len(normVals) - 1; i >= 0; i-- {
+			if _, ok := sets[normVals[i]]; ok {
+				normVals[i] = normVals[len(normVals)-1]
+				normVals = normVals[:len(normVals)-1]
+			} else {
+				sets[normVals[i]] = struct{}{}
+			}
+		}
+		normValsArr = append(normValsArr, normVals)
+	}
+
+	arrFieldCounter := make([]int, len(index.arrFieldsIndexes))
+	done := false
+
+	return func() (core.IndexDataStoreKey, bool) {
+		if done {
+			return core.IndexDataStoreKey{}, false
+		}
+
+		resultKey := core.IndexDataStoreKey{
+			CollectionID: key.CollectionID,
+			IndexID:      key.IndexID,
+			Fields:       make([]core.IndexedField, len(key.Fields)),
+		}
+		copy(resultKey.Fields, key.Fields)
+
+		for i, counter := range arrFieldCounter {
+			field := &resultKey.Fields[index.arrFieldsIndexes[i]]
+			field.Value = normValsArr[i][counter]
+		}
+
+		for i := len(arrFieldCounter) - 1; i >= 0; i-- {
+			arrFieldCounter[i]++
+			if arrFieldCounter[i] < len(normValsArr[i]) {
+				break
+			}
+			arrFieldCounter[i] = 0
+			if i == 0 {
+				done = true
+			}
+		}
+
+		return resultKey, true
+	}, nil
+}
+
+func (index *collectionArrayBaseIndex) getAllKeys(
+	doc *client.Document,
+	appendDocID bool,
+) ([]core.IndexDataStoreKey, error) {
+	getNextOldKey, err := index.newIndexKeyGenerator(doc, appendDocID)
+	if err != nil {
+		return nil, err
+	}
+	keys := make([]core.IndexDataStoreKey, 0)
+	for {
+		key, ok := getNextOldKey()
+		if !ok {
+			break
+		}
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+func (index *collectionArrayBaseIndex) deleteRetiredKeysAndReturnNew(
+	ctx context.Context,
+	txn datastore.Txn,
+	oldDoc *client.Document,
+	newDoc *client.Document,
+	appendDocID bool,
+) ([]core.IndexDataStoreKey, error) {
+	oldKeys, err := index.getAllKeys(oldDoc, appendDocID)
+	if err != nil {
+		return nil, err
+	}
+	newKeys, err := index.getAllKeys(newDoc, appendDocID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, oldKey := range oldKeys {
+		isFound := false
+		for i := len(newKeys) - 1; i >= 0; i-- {
+			if oldKey.IsEqual(newKeys[i]) {
+				newKeys[i] = newKeys[len(newKeys)-1]
+				newKeys = newKeys[:len(newKeys)-1]
+				isFound = true
+				break
+			}
+		}
+		if !isFound {
+			err = index.deleteIndexKey(ctx, txn, oldKey)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return newKeys, nil
+}
+
+type collectionArrayIndex struct {
+	collectionArrayBaseIndex
+}
+
+var _ CollectionIndex = (*collectionArrayIndex)(nil)
+
+func newCollectionArrayIndex(base collectionBaseIndex) *collectionArrayIndex {
+	return &collectionArrayIndex{collectionArrayBaseIndex: newCollectionArrayBaseIndex(base)}
+}
+
+// Save indexes a document by storing the indexed field value.
+func (index *collectionArrayIndex) Save(
```

Comment:
question: Why does `Save` not retire the old keys? If it does, please include this in the function level documentation

Reply:
I've updated the documentation for the interface.